src/Tools/Source/RunTests/AssemblyScheduler.cs

```diff
@@ -2,359 +2,314 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
+using System.Collections;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Diagnostics.Contracts;
 using System.IO;
 using System.Linq;
 using System.Reflection;
 using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
 using System.Reflection.PortableExecutable;
-using System.Text;
+using System.Runtime.InteropServices;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.Common;
 
 namespace RunTests
 {
-    internal readonly struct AssemblyInfo
+    internal record struct WorkItemInfo(ImmutableSortedDictionary<AssemblyInfo, ImmutableArray<TestMethodInfo>> Filters, int PartitionIndex)
     {
-        internal PartitionInfo PartitionInfo { get; }
-        internal string TargetFramework { get; }
-        internal string Architecture { get; }
-
-        internal string AssemblyPath => PartitionInfo.AssemblyPath;
-        internal string AssemblyName => Path.GetFileName(PartitionInfo.AssemblyPath);
-        internal string DisplayName => PartitionInfo.DisplayName;
-
-        internal AssemblyInfo(PartitionInfo partitionInfo, string targetFramework, string architecture)
-        {
-            PartitionInfo = partitionInfo;
-            TargetFramework = targetFramework;
-            Architecture = architecture;
-        }
+        internal string DisplayName => $"{string.Join("_", Filters.Keys.Select(a => Path.GetFileNameWithoutExtension(a.AssemblyName)))}_{PartitionIndex}";
     }
 
-    internal readonly struct PartitionInfo
+    internal sealed class AssemblyScheduler
     {
-        internal int? AssemblyPartitionId { get; }
-        internal string AssemblyPath { get; }
-        internal string DisplayName { get; }
+        /// <summary>
+        /// Our execution time limit is 2m30s.  We really want to run tests under 5 minutes, but we need to limit test execution time
```

Comment:
What imposes this limit?

Reply:
THis is a derived limit - we want to run our tests in under 5min generally, but to account for test discovery and helix work item setup, the work items really should be running in under 2:30.  We then build partitions based on historical execution time to hit that limit

Reply:
Can we be clearer on that? As written it sounds like there is a hard limit here.